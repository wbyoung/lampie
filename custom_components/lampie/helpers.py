"""Lampie helpers."""

from __future__ import annotations

from collections.abc import AsyncGenerator
from contextlib import asynccontextmanager
import logging

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.device import async_device_info_to_link_from_entity
from homeassistant.util import slugify

from .const import CONF_PRIORITY, DOMAIN
from .types import LampieConfigEntry, LampieConfigEntryRuntimeData

_LOGGER = logging.getLogger(__name__)


def is_primary_for_switch(entry: LampieConfigEntry, switch_id: str) -> bool:
    priorities = entry.data.get(CONF_PRIORITY, {}).get(switch_id)

    if not priorities:
        return True
    return bool(priorities[0] == slugify(entry.title))


def lookup_device_info(
    hass: HomeAssistant, entry: LampieConfigEntry, switch_id: str
) -> dr.DeviceInfo:
    device_info = async_device_info_to_link_from_entity(hass, switch_id)

    if not device_info:
        _LOGGER.warning(
            "skipping creation of sensors for %s on %s because an associated "
            "device could not be found",
            switch_id,
            entry.title,
        )

    return device_info


def get_other_entries(
    config_entry: ConfigEntry, *, hass: HomeAssistant
) -> list[ConfigEntry]:
    return [
        entry
        for entry in hass.config_entries.async_entries(DOMAIN)
        if entry is not config_entry
    ]


@asynccontextmanager
async def auto_reload_disabled(  # noqa: RUF029
    other_entries: list[ConfigEntry],
) -> AsyncGenerator[None]:
    targets: list[LampieConfigEntryRuntimeData] = [
        entry.runtime_data for entry in other_entries if hasattr(entry, "runtime_data")
    ]

    for runtime_data in targets:
        runtime_data.auto_reload_enabled = False

    try:
        yield
    finally:
        for runtime_data in targets:
            runtime_data.auto_reload_enabled = True


@asynccontextmanager
async def unloaded(
    other_entries: list[ConfigEntry],
    *,
    hass: HomeAssistant,
) -> AsyncGenerator[None]:
    other_entries = [  # filter to only the loaded entries
        entry for entry in other_entries if entry.state is ConfigEntryState.LOADED
    ]
    for entry in other_entries:
        await hass.config_entries.async_unload(entry.entry_id)

    yield

    for entry in other_entries:
        assert entry.state is ConfigEntryState.NOT_LOADED
        await hass.config_entries.async_setup(entry.entry_id)
