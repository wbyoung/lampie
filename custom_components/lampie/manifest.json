{
  "domain": "lampie",
  "name": "Lampie",
  "codeowners": [
    "@wbyoung"
  ],
  "config_flow": true,
  "dependencies": ["mqtt"],
  "documentation": "https://github.com/wbyoung/lampie",
  "homekit": {},
  "iot_class": "calculated",
  "issue_tracker": "https://github.com/wbyoung/lampie/issues",
  "requirements": [],
  "ssdp": [],
  "version": "0.1.0",
  "zeroconf": []
}
