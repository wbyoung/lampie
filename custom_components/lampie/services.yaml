activate:
  fields:
    notification:
      required: true
      selector:
        text:
    leds:
      required: false
      selector:
        object:
override:
  fields:
    entity_id:
      required: true
      selector:
        entity:
          multiple: true
          filter:
            - integration: zha
              domain:
                - light
                - fan
            - integration: mqtt
              domain:
                - light
                - fan
    leds:
      required: true
      selector:
        object:
    name:
      required: false
      selector:
        text:
