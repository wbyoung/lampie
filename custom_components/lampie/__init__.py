"""The Lampie integration."""

from __future__ import annotations

import logging

from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.event import async_track_entity_registry_updated_event
from homeassistant.helpers.typing import ConfigType

from .const import CONF_SWITCH_ENTITIES, DOMAIN
from .coordinator import LampieUpdateCoordinator
from .orchestrator import LampieOrchestrator
from .services import async_setup_services
from .types import LampieConfigEntry, LampieConfigEntryRuntimeData

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [Platform.SENSOR, Platform.SWITCH]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


async def async_setup(  # noqa: RUF029
    hass: HomeAssistant,
    config: ConfigType,  # noqa: ARG001
) -> bool:
    """Set up Lampie services.

    Returns:
        If the setup was successful.
    """
    async_setup_services(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: LampieConfigEntry) -> bool:
    """Set up Lampie from a config entry.

    Returns:
        If the setup was successful.
    """
    _LOGGER.debug("setup %s with config:%s", entry.title, entry.data)

    if DOMAIN not in hass.data:
        orchestrator = LampieOrchestrator(hass)
        hass.data[DOMAIN] = orchestrator

        await orchestrator.setup()

    coordinator = LampieUpdateCoordinator(hass, entry)
    orchestrator = hass.data[DOMAIN]
    await orchestrator.add_coordinator(coordinator)
    entry.runtime_data = LampieConfigEntryRuntimeData(
        orchestrator=orchestrator,
        coordinator=coordinator,
    )

    entry.async_on_unload(
        async_track_entity_registry_updated_event(
            hass,
            entry.data[CONF_SWITCH_ENTITIES],
            coordinator.async_handle_switch_id_change,
        ),
    )

    await coordinator.async_config_entry_first_refresh()
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
    return True


async def async_unload_entry(hass: HomeAssistant, entry: LampieConfigEntry) -> bool:
    """Unload a config entry.

    Returns:
        If the unload was successful.
    """
    coordinator: LampieUpdateCoordinator = entry.runtime_data.coordinator
    unload_ok: bool = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)

    if orchestrator := entry.runtime_data.orchestrator:
        await orchestrator.remove_coordinator(coordinator)

        if orchestrator.teardown() and orchestrator == hass.data.get(DOMAIN):
            hass.data.pop(DOMAIN)

    return unload_ok


async def _async_update_listener(
    hass: HomeAssistant,
    entry: LampieConfigEntry,
) -> None:
    """Handle options update."""
    runtime_data: LampieConfigEntryRuntimeData = entry.runtime_data

    if runtime_data.auto_reload_enabled:
        await hass.config_entries.async_reload(entry.entry_id)
