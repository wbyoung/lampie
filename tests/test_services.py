"""Test services for the Lampie integration."""

import re
from typing import Any

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import entity_registry as er
import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    async_mock_service,
)
from syrupy.assertion import SnapshotAssertion
import voluptuous as vol

from custom_components.lampie.const import (
    ATTR_ENTITY_ID,
    ATTR_LED_CONFIG,
    ATTR_NOTIFICATION,
    DOMAIN,
)
from custom_components.lampie.services import (
    SERVICE_NAME_ACTIVATE,
    SERVICE_NAME_OVERRIDE,
)


@pytest.mark.usefixtures("init_integration")
def test_has_services(
    hass: HomeAssistant,
) -> None:
    """Test the existence of the Lampie Service."""
    assert hass.services.has_service(DOMAIN, SERVICE_NAME_ACTIVATE)
    assert hass.services.has_service(DOMAIN, SERVICE_NAME_OVERRIDE)


@pytest.mark.usefixtures("switch", "init_integration")
@pytest.mark.parametrize(
    "attrs",
    [
        {ATTR_NOTIFICATION: "doors_open"},
        {ATTR_NOTIFICATION: "doors_open", ATTR_LED_CONFIG: "green"},
        {
            ATTR_NOTIFICATION: "doors_open",
            ATTR_LED_CONFIG: [
                "green",
                "orange",
                "blue",
                "yellow",
                "pink",
                "red",
                {
                    "color": "cyan",
                    "effect": "fast_blink",
                },
            ],
        },
    ],
    ids=["notification_only", "led_color", "full_led_config"],
)
async def test_activate_service(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    switch: er.RegistryEntry,
    attrs: dict[str, Any],
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
):
    """Test `activate` service call."""
    cluster_commands = async_mock_service(hass, "zha", "issue_zigbee_cluster_command")

    await hass.services.async_call(
        DOMAIN,
        SERVICE_NAME_ACTIVATE,
        attrs,
        blocking=True,
    )

    assert cluster_commands == snapshot(name="zha_cluster_commands")

    for entity in entity_registry.entities.get_entries_for_config_entry_id(
        config_entry.entry_id
    ):
        assert hass.states.get(entity.entity_id) == snapshot(name=entity.entity_id)


@pytest.mark.usefixtures("init_integration")
@pytest.mark.parametrize(
    "attrs",
    [
        {ATTR_ENTITY_ID: "light.kitchen", ATTR_LED_CONFIG: "green"},
        {
            ATTR_ENTITY_ID: "light.kitchen",
            ATTR_LED_CONFIG: [
                "green",
                "orange",
                "blue",
                "yellow",
                "pink",
                "red",
                {
                    "color": "cyan",
                    "effect": "fast_blink",
                },
            ],
        },
    ],
    ids=["led_color", "full_led_config"],
)
async def test_override_service(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    switch: er.RegistryEntry,
    attrs: dict[str, Any],
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
):
    """Test `override` service call."""
    cluster_commands = async_mock_service(hass, "zha", "issue_zigbee_cluster_command")

    await hass.services.async_call(
        DOMAIN,
        SERVICE_NAME_OVERRIDE,
        attrs,
        blocking=True,
    )

    assert cluster_commands == snapshot(name="zha_cluster_commands")

    for entity in entity_registry.entities.get_entries_for_config_entry_id(
        config_entry.entry_id
    ):
        assert hass.states.get(entity.entity_id) == snapshot(name=entity.entity_id)


@pytest.mark.usefixtures("init_integration")
@pytest.mark.parametrize(
    ("service_data", "error", "error_message"),
    [
        ({}, vol.er.Error, "required key not provided .+"),
        (
            {"notification": "doors_open", "leds": "pinkish"},
            vol.er.Error,
            ".+ not a valid color.+name",
        ),
        (
            {"notification": "incorrect_slug"},
            ServiceValidationError,
            "Invalid notification .+",
        ),
    ],
)
async def test_activate_service_validation(
    hass: HomeAssistant,
    service_data: dict[str, str],
    error: type[Exception],
    error_message: str,
) -> None:
    """Test the `activate` service validation."""
    with pytest.raises(error) as exc:
        await hass.services.async_call(
            DOMAIN,
            SERVICE_NAME_ACTIVATE,
            service_data,
            blocking=True,
        )
    assert re.match(error_message, str(exc.value))


@pytest.mark.usefixtures("init_integration")
@pytest.mark.parametrize(
    ("service_data", "error", "error_message"),
    [
        ({"entity_id": "light.kitchen"}, vol.er.Error, "required key not provided .+"),
        ({"leds": "green"}, vol.er.Error, "required key not provided .+"),
        (
            {"entity_id": "light.kitchen", "leds": "pinkish"},
            vol.er.Error,
            ".+ not a valid color.+name",
        ),
    ],
)
async def test_override_service_validation(
    hass: HomeAssistant,
    service_data: dict[str, str],
    error: type[Exception],
    error_message: str,
) -> None:
    """Test the `override` service validation."""
    with pytest.raises(error) as exc:
        await hass.services.async_call(
            DOMAIN,
            SERVICE_NAME_OVERRIDE,
            service_data,
            blocking=True,
        )
    assert re.match(error_message, str(exc.value))
