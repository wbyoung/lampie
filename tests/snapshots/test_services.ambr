# serializer version: 1
# name: test_activate_service[full_led_config][sensor.kitchen_effect_brightness]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect brightness',
      'icon': 'mdi:brightness-percent',
      'individual': list([
        100.0,
        100.0,
        100.0,
        100.0,
        100.0,
        100.0,
        100.0,
      ]),
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_brightness',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '100.0',
  })
# ---
# name: test_activate_service[full_led_config][sensor.kitchen_effect_color]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect color',
      'icon': 'mdi:palette',
      'individual': list([
        'green',
        'orange',
        'blue',
        'yellow',
        'pink',
        'red',
        'cyan',
      ]),
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_color',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'multi',
  })
# ---
# name: test_activate_service[full_led_config][sensor.kitchen_effect_duration]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect duration',
      'icon': 'mdi:timer',
      'individual': list([
        None,
        None,
        None,
        None,
        None,
        None,
        None,
      ]),
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTime.SECONDS: 's'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_duration',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_activate_service[full_led_config][sensor.kitchen_effect_type]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect type',
      'icon': 'mdi:star-four-points-box',
      'individual': list([
        'solid',
        'solid',
        'solid',
        'solid',
        'solid',
        'solid',
        'fast_blink',
      ]),
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_type',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'multi',
  })
# ---
# name: test_activate_service[full_led_config][sensor.kitchen_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'doors_open[custom]',
  })
# ---
# name: test_activate_service[full_led_config][switch.doors_open_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Doors Open Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'switch.doors_open_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_activate_service[full_led_config][zha_cluster_commands]
  list([
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 90,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 0,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 25,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 1,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 170,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 2,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 45,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 3,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 230,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 4,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 0,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 5,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 130,
          'led_duration': 255,
          'led_effect': 2,
          'led_level': 100,
          'led_number': 6,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
  ])
# ---
# name: test_activate_service[led_color][sensor.kitchen_effect_brightness]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect brightness',
      'icon': 'mdi:brightness-percent',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_brightness',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '100.0',
  })
# ---
# name: test_activate_service[led_color][sensor.kitchen_effect_color]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect color',
      'icon': 'mdi:palette',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_color',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'green',
  })
# ---
# name: test_activate_service[led_color][sensor.kitchen_effect_duration]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect duration',
      'icon': 'mdi:timer',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTime.SECONDS: 's'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_duration',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_activate_service[led_color][sensor.kitchen_effect_type]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect type',
      'icon': 'mdi:star-four-points-box',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_type',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'solid',
  })
# ---
# name: test_activate_service[led_color][sensor.kitchen_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'doors_open[custom]',
  })
# ---
# name: test_activate_service[led_color][switch.doors_open_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Doors Open Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'switch.doors_open_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_activate_service[led_color][zha_cluster_commands]
  list([
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 1,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 90,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
  ])
# ---
# name: test_activate_service[notification_only][sensor.kitchen_effect_brightness]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect brightness',
      'icon': 'mdi:brightness-percent',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_brightness',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '100.0',
  })
# ---
# name: test_activate_service[notification_only][sensor.kitchen_effect_color]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect color',
      'icon': 'mdi:palette',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_color',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'red',
  })
# ---
# name: test_activate_service[notification_only][sensor.kitchen_effect_duration]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect duration',
      'icon': 'mdi:timer',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTime.SECONDS: 's'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_duration',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_activate_service[notification_only][sensor.kitchen_effect_type]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Effect type',
      'icon': 'mdi:star-four-points-box',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_effect_type',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'open_close',
  })
# ---
# name: test_activate_service[notification_only][sensor.kitchen_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Kitchen Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.kitchen_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'doors_open',
  })
# ---
# name: test_activate_service[notification_only][switch.doors_open_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Doors Open Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'switch.doors_open_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_activate_service[notification_only][zha_cluster_commands]
  list([
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 1,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 0,
          'led_duration': 255,
          'led_effect': 6,
          'led_level': 100,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
  ])
# ---
# name: test_override_service[full_led_config][switch.doors_open_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Doors Open Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'switch.doors_open_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_override_service[full_led_config][zha_cluster_commands]
  list([
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 90,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 0,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 25,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 1,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 170,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 2,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 45,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 3,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 230,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 4,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 0,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
          'led_number': 5,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 3,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 130,
          'led_duration': 255,
          'led_effect': 2,
          'led_level': 100,
          'led_number': 6,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
  ])
# ---
# name: test_override_service[led_color][switch.doors_open_notification]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Doors Open Notification',
      'icon': 'mdi:circle-box',
    }),
    'context': <ANY>,
    'entity_id': 'switch.doors_open_notification',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_override_service[led_color][zha_cluster_commands]
  list([
    ServiceCallSnapshot({
      'context': <ANY>,
      'data': ReadOnlyDict({
        'cluster_id': 64561,
        'cluster_type': 'in',
        'command': 1,
        'command_type': 'server',
        'endpoint_id': 1,
        'ieee': 'mock-ieee:kitchen',
        'manufacturer': 4655,
        'params': dict({
          'led_color': 90,
          'led_duration': 255,
          'led_effect': 1,
          'led_level': 100,
        }),
      }),
      'domain': 'zha',
      'hass': <ANY>,
      'return_response': False,
      'service': 'issue_zigbee_cluster_command',
    }),
  ])
# ---
